### PR TITLE
Optimize `<xstring>` memmove function

### PR DESCRIPTION
**Pull request by gAlfonso-bit**
Optimized memmove

### CODE REVIEW

**Comment by fsb4000** [comment]
CI errors:

```
D:\build\out\inc\cstring(26,13): error: no member named 'memccpy' in the global namespace

using _CSTD memccpy;

      ~~~~~ ^

D:\build\out\inc\cstring(38,13): error: no member named 'strdup' in the global namespace; did you mean '_strdup'?

using _CSTD strdup;

      ~~~~~ ^~~~~~

            _strdup

C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\ucrt\string.h(148,38): note: '_strdup' declared here

_ACRTIMP _CRTALLOCATOR char* __cdecl _strdup(

                                     ^

2 errors generated.
```

P.S. I doubt your PR will be accepted because `Non-goal: Adding non-Standard extensions.`
But I am not a maintainer of this repository so I could be wrong.
You will have a better chance that your PR will be accepted if you solve some existing issue.  
Check out https://github.com/microsoft/STL/issues

**Comment by miscco** [comment]
I do not see any functional code change in product code. Please revert purely aesthetical changes

**Comment by gAlfonso-bit** [comment]
> I do not see any functional code change in product code. Please revert purely aesthetical changes

Done

File: stl/inc/xstring
Comment (cbezault) [inline stl/inc/xstring:114]: Does this actually provide superior codegen?

Comment (AdamBucior) [reply stl/inc/xstring:114]: This code is only executed at compile-time.

File: stl/inc/cstring
Comment (StephanTLavavej) [inline stl/inc/cstring:42]: Speaking as a maintainer, I don't believe that we should merge this change (or the corresponding test change in `tests/tr1/tests/cstring/test.cpp`). Microsoft's `::strnlen` is a non-Standard extension, and while we can't get rid of it, we're trying to avoid adding non-Standard extensions going forward (particularly those that involve names that aren't reserved/`_Ugly` as reserved names indicate that something inherently non-Standard is happening).

For precedent/guidance, the Standard itself strongly suggests that `namespace std` should be kept "clean". WG21-N4892 16.4.2.3 [headers]/10 says:

> Annex K of the C standard describes a large number of functions, with associated types and macros, which "promote safer, more secure programming" than many of the traditional C library functions. The names of the functions have a suffix of `_s`; most of them provide the same service as the C library function with the unsuffixed name, but generally take an additional argument whose value is the size of the result array. If any C++ header is included, it is implementation-defined whether any of these names is declared in the global namespace. (None of them is declared in namespace `std`.)

And this applies to [C11](http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1570.pdf) K.3.7.4.4 "The `strnlen_s` function". If this Annex K function shouldn't be in `std`, plain `::strnlen` shouldn't either.

Additionally, even in the distant past when the STL added non-Standard machinery, we did so in the `stdext` namespace (e.g. `stdext::hash_map`, `stdext::checked_array_iterator`) in order to make it clear that `std` is truly Standard. The only major exceptions to this policy that I'm aware of are the TR1-era names `mersenne_twister` etc. in `<random>` which are historical relics (and which we want to remove, but binary compatibility makes it difficult).

Ultimately, this change wouldn't make it easier for users to write portable code, so I believe this falls under the project's [non-goals](https://github.com/microsoft/STL#non-goals). (I do appreciate the attempt to increase consistency, though!)

File: stl/inc/xstring
Comment (StephanTLavavej) [reply stl/inc/xstring:114]: I wrote this code in @miscco's #491 (in commit https://github.com/microsoft/STL/pull/491/commits/fd291a8991282cead9b160bb71e024758d7b9e71 ). Looking at it now, I can't remember why I wrote it as computing `[_Count - 1 - _Idx]`, and I agree that the proposed change is a clarity improvement (even if it doesn't affect runtime perf or compiletime throughput). There's no need for the "loop forward" and "loop backward" cases to have exactly the same `for` loop, there's no concern about `size_t` wrapping around (the new code cleanly stops when `_Idx` falls to `0`, and does nothing if it starts at `0`), and no danger of accessing `[-1]`. I think I simply missed how this could be clearly written with `[_Idx - 1]`.

I believe that we should merge this change (with the other one dropped, and the PR title appropriately changed). Thanks for finding and improving this weird bit of code that I wrote! :smile_cat:

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]

File: stl/inc/xstring
Comment (gAlfonso-bit) [reply stl/inc/xstring:114]: Fixed

**Review by fsb4000 — APPROVED** [review]

**Review by CaseyCarter — APPROVED** [review]

**Review by sam20908 — APPROVED** [review]

**Review by StephanTLavavej — APPROVED** [review]

**Review by CaseyCarter — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks for this simplification, and congratulations on your first microsoft/STL commit! :rocket: :smiley_cat: This will ship in VS 2022 17.1 Preview 1.

A note for the future: when pushing changes after reviewers have approved, it's a good idea to add a comment to the PR mentioning that you've done so, so the reviewers know to take another look. (Some changes that seem obviously desirable can be inconsistent with a coding convention, or can have surprising interactions with language complexity, etc. - which is why we're so careful about getting all changes to be seen by multiple reviewers.) Here, Casey and I had approved and moved this to Ready To Merge, and I wasn't looking closely for unannounced changes, so your last commit caught me by surprise. Fortunately Casey noticed and re-approved, so all was well (and I'm also fine with the change). This also happened before I mirrored the GitHub PR to the MSVC-internal repo, so no changes were dropped on the floor. I've now realized that we need to document these expectations better, so I'll add this to the Contribution Guidelines document that I need to write up.

**Pull request merged**
